include/ethyl/utils.hpp

```diff
@@ -38,23 +43,72 @@ namespace utils
     std::string_view trimPrefix(std::string_view src, std::string_view prefix);
     std::string_view trimLeadingZeros(std::string_view src);
 
-    std::vector<unsigned char>    fromHexString(std::string_view hexStr);
-    uint64_t                      fromHexStringToUint64(std::string_view hexStr);
-    std::array<unsigned char, 32> fromHexString32Byte(std::string_view hexStr);
+    using oxenc::basic_char;
+    template <basic_char Char = unsigned char>
+    std::vector<Char> fromHexString(std::string_view hexStr) {
+        hexStr = trimPrefix(hexStr, "0x");
 
-    std::array<unsigned char, 32> hash(std::string in);
+        if (!oxenc::is_hex(hexStr))
+            throw std::invalid_argument{"input string is not hex"};
 
-    std::string getFunctionSignature(const std::string& function);
+        std::vector<Char> result;
+        result.reserve(oxenc::from_hex_size(hexStr.size()));
+        oxenc::from_hex(hexStr.begin(), hexStr.end(), std::back_inserter(result));
+        return result;
+    }
+    extern template std::vector<char> fromHexString<char>(std::string_view);
+    extern template std::vector<unsigned char> fromHexString<unsigned char>(std::string_view);
 
-    std::string padToNBytes(const std::string& input, size_t byteCount, PaddingDirection direction = PaddingDirection::LEFT);
-    std::string padTo8Bytes(const std::string& input, PaddingDirection direction = PaddingDirection::LEFT);
-    std::string padTo32Bytes(const std::string& input, PaddingDirection direction = PaddingDirection::LEFT);
+    template <basic_char Char = unsigned char>
+    std::array<Char, 32> fromHexString32Byte(std::string_view hexStr) {
+        hexStr = trimPrefix(hexStr, "0x");
 
-    std::vector<unsigned char> intToBytes(uint64_t num);
+        if (!oxenc::is_hex(hexStr) || hexStr.size() != 64) {
+            throw std::invalid_argument("Input string length should be 64 hex characters for 32 bytes");
+        }
+
+        std::array<Char, 32> bytesArr;
+        oxenc::from_hex(hexStr.begin(), hexStr.end(), bytesArr.begin());
 
-    std::vector<unsigned char> removeLeadingZeros(std::vector<unsigned char> vec);
+        return bytesArr;
+    }
+    extern template std::array<char, 32> fromHexString32Byte<char>(std::string_view);
+    extern template std::array<unsigned char, 32> fromHexString32Byte<unsigned char>(std::string_view);
+
+
+    uint64_t fromHexStringToUint64(std::string_view hexStr);
+
+    std::string padToNBytes(std::string_view hex_input, size_t bytes, PaddingDirection direction = PaddingDirection::LEFT);
+    inline std::string padTo8Bytes(std::string_view hex_input, PaddingDirection direction = PaddingDirection::LEFT) {
+        return padToNBytes(hex_input, 32, direction);
```

Comment:
pads to 32 bytes when it should be padding to 8 bytes. Looks like the other function has it swapped also

Reply:
Fixed.